blog: Podman and Docker Rootless in DDEV, fixes #453

stasadev opened the pull request:
## The Issue

- #453

## How This PR Solves The Issue

Adds a blog.

## Manual Testing Instructions

https://pr-476.ddev-com-fork-previews.pages.dev/blog/podman-and-docker-rootless/

## Automated Testing Overview

<!-- Please describe the tests introduced by this PR, or explain why no tests are needed. -->

## Related Issue Link(s)

## Release/Deployment Notes

<!-- Does this affect anything else or have ramifications for other code? Does anything have to be done on deployment? -->

rfay commented on src/content/blog/podman-and-docker-rootless.md:
```diff
@@ -0,0 +1,481 @@
+- **Removed legacy docker-compose features** ([#7642](https://github.com/ddev/ddev/pull/7642)): Podman refuses to work with deprecated `links` and `external_links` directives in `docker-compose` files. We removed these legacy features and modernized DDEV's compose file generation.
+- **Added Podman and Docker Rootless support** ([#7702](https://github.com/ddev/ddev/pull/7702)): DDEV now detects and supports Podman (rootful and rootless) and Docker Rootless. We added handling for Podman-specific limitations and enabled rootless environments to work without root privileges.
+
+These changes enabled Podman and Docker Rootless support.
```
An amazing journey, congratulations!

rfay commented on src/content/blog/podman-and-docker-rootless.md:
```diff
@@ -0,0 +1,481 @@
+- Container processes cannot access root-owned files
+- Reduced attack surface if a container is compromised
+
+Podman is rootless by default. Docker Rootless requires special setup.
```
It is probably worth explaining here that DDEV's traditional support has had its containers run as an unprivileged user, limiting the attack space. DDEV's built-in containers do not use root privileges (although docker itself does need root by default)

rfay commented on src/content/blog/podman-and-docker-rootless.md:
```diff
@@ -0,0 +1,481 @@
+
+**With Podman Desktop GUI**:
+
+- [macOS](https://podman-desktop.io/docs/installation/macos-install)
```
Interesting that they discourage use of homebrew.

rfay commented on src/content/blog/podman-and-docker-rootless.md:
```diff
@@ -0,0 +1,481 @@
+
+After installing Podman, enable the API socket so DDEV can connect.
+
+### Podman Rootless on Linux
```
Should this say "Linux and WSL2" ?

rfay commented on src/content/blog/podman-and-docker-rootless.md:
````diff
@@ -0,0 +1,481 @@
+   sudo podman info --format '{{.Host.RemoteSocket.Path}}'
+   ```
+
+7. Configure Docker to use Podman. Choose one of these methods:
````
This statement will be confusing. I assume this means "Configure `docker` CLI to connect to Podman instead of Docker Daemon."

rfay commented on src/content/blog/podman-and-docker-rootless.md:
````diff
@@ -0,0 +1,481 @@
+   **Option B: Set environment variable in your shell profile** (`~/.bashrc` or `~/.zshrc`):
+
+   ```bash
+   export DOCKER_HOST=unix:///var/run/podman/podman.sock
````
I wonder if we should not mention this technique. It seems like it will be fragile and little used.

rfay commented on src/content/blog/podman-and-docker-rootless.md:
```diff
@@ -0,0 +1,481 @@
+   - [Mounting a volume with rootless always assigns ownership to root](https://github.com/moby/moby/issues/45919)
+   - [Add ability to mount volume as user other than root](https://github.com/moby/moby/issues/2259)
+
+   The `root` user inside the container maps to your host user, but many services will not run as root:
```
It's interesting that this is the same classic problem we've had with Docker Desktop for Linux and for a time with virtiofs.

rfay commented on src/content/blog/podman-and-docker-rootless.md:
````diff
@@ -0,0 +1,481 @@
+   ddev start
+   ```
+
+## Which Runtime Should You Choose?
````
Great section. However, it should encourage standard Docker setup first, with link to the docs, as it's by far the best for local development.

rfay commented:
This and docs probably need a compatibility table showing all the options and what works etc.

rfay commented on src/content/blog/podman-and-docker-rootless.md:
```diff
@@ -0,0 +1,481 @@
+
+### Podman Installation Options
+
+Podman can be installed with or without a GUI:
```
I think we probably need a little more than this. I missed both `podman init` and `podman start`. Maybe one of those collapsed sections with basic instructions on macOS and Linux? The majority of our users are macOS, but probably the majority of interested users are on Linux.

stasadev replied:
I'll check the formatting options for Markdown here.

stasadev commented on src/content/blog/podman-and-docker-rootless.md:
````diff
@@ -0,0 +1,593 @@
+
+```bash
+export DOCKER_HOST=unix://$(podman info --format '{{.Host.RemoteSocket.Path}}')
+```
````
I didn't test macOS, hoping for possible improvements or tips from @rfay in this section.

rfay commented on src/content/blog/podman-and-docker-rootless.md:
```diff
@@ -0,0 +1,598 @@
+
+All of these work with DDEV. The main reason to choose Podman specifically is if your organization forbids Docker entirely or if you want rootless operation by default.
+
+### Why Choose Rootless?
```
Great section. Maybe "Why choose Podman" right after it. Or before it.

rfay commented on src/content/blog/podman-and-docker-rootless.md:
```diff
@@ -0,0 +1,598 @@
+
+### Why Choose Rootless?
+
+Running containers without root privileges is more secure. Traditional Docker and rootful Podman need elevated privileges, which creates security risks in corporate environments where strict security policies apply.
```
```suggestion
Although DDEV's use of all Docker providers is quite secure, and we run containers as normal users with limited privileges, the rootless approaches to Docker and Podman actually run the *docker daemon* without root privileges, closing additional attack surface. Traditional Docker and rootful Podman daemons need elevated privileges, which creates security risks in corporate environments where strict security policies apply. (Note that DDEV is targeted at local development, where there are few risks of specialized attacks using this vector anyway.)
```

rfay commented on src/content/blog/podman-and-docker-rootless.md:
```diff
@@ -0,0 +1,598 @@
+
+### Podman Rootless Performance Optimization
+
+Podman Rootless is slower than Docker. See these resources:
```
```suggestion
Podman Rootless is significantly slower than Docker. See these resources:
```

stasadev commented:
I edited the blog:

- Removed DOCKER_HOST
- Added explanation why to install `docker` CLI
- Removed all the instructions from Podman rootful, but the section remained to show the difference
- Simplified instructions
- Added checks for already applied settings based on what was done on Fedora

I'm not sure that macOS instructions are correct.

stasadev commented on src/content/blog/podman-and-docker-rootless.md:
````diff
@@ -0,0 +1,519 @@
+podman system reset
+```
+
+This removes all existing containers, images, and volumes (similar to `docker system prune -a`).
````
I tested it on Arch Linux, and `docker ps` worked fine without reboot after `podman system reset`.

> but after reboot I could

Maybe this command can help to do it without reboot:

```bash
systemctl --user restart podman.socket
```

stasadev commented:
TODO: review the blog, and pull it after:

- #526

rfay commented on src/content/blog/podman-and-docker-rootless.md:
```diff
@@ -0,0 +1,568 @@
+  - Guides
+---
+
+**TL;DR**: DDEV now supports Podman and Docker Rootless in [DDEV HEAD](https://docs.ddev.com/en/stable/developers/building-contributing/#testing-latest-commits-on-head) (coming in v1.25.0). Jump to setup instructions: [Linux/WSL2](#key-aim-linux-and-wsl2-users) · [macOS](#macos) · [Windows](#windows)
```
Let's expand this.


```suggestion
**TL;DR**: DDEV now supports Podman and Docker Rootless in [DDEV HEAD](https://docs.ddev.com/en/stable/developers/building-contributing/#testing-latest-commits-on-head) (coming in v1.25.0).  Podman and Docker Rootless are a bit more trouble than the [recommended normal traditional docker providers](https://docs.ddev.com/en/stable/users/install/docker-installation/) and have some serious trade-offs. On macOS you can't use the nomal default ports 80 and 443. On Linux Docker Rootless you can't bind-mount directories, so the entire project has to mutagen-synced.

Jump to setup instructions: [Linux/WSL2](#key-aim-linux-and-wsl2-users) · [macOS](#macos) · [Windows](#windows)
```

stasadev commented on src/content/blog/podman-and-docker-rootless.md:
```diff
@@ -0,0 +1,568 @@
+
+3. Proceed with [DDEV installation](https://docs.ddev.com/en/stable/users/install/ddev-installation/).
+
+4. Handle privileged ports (<1024):
```
Okay.

> I do wonder why our automatic switching to ephemeral ports doesn't work in this situation.

Since the ephemeral ports don't check the actual bind, the 80/443 ports are available for use.

rfay commented on src/content/blog/podman-and-docker-rootless.md:
```diff
@@ -0,0 +1,593 @@
+   systemctl --user enable --now podman.socket
+
+   # Socket path for rootless
+   ls $XDG_RUNTIME_DIR/podman/podman.sock
```
I think the "compoare to rootless" will just confuse people and they'll try to follow these instructions.

I don't think XDG_RUNTIME_DIR is there on Ubuntu btw

rfay commented on src/content/blog/podman-and-docker-rootless.md:
````diff
@@ -0,0 +1,593 @@
+
+   ```bash
+   # Enable rootful socket (requires sudo)
+   sudo systemctl enable --now podman.socket
````
With this approach we also need 
```
export DOCKER_HOST=unix:///var/run/podman/podman.sock
```

or instructions to create a context

stasadev commented on src/content/blog/podman-and-docker-rootless.md:
```diff
@@ -0,0 +1,593 @@
+
+   # Switch to the context
+   docker context use rootless
+
```
Maybe, but running it doesn't break anything, because I check for the context before creating it.

I didn't have any context created on Arch-based Linux (because I didn't run the script; it's not intended for Arch).
